contrib/parseq-batching/src/main/java/com/linkedin/parseq/batching/BatchingStrategy.java

```diff
@@ -0,0 +1,282 @@
+      return result;
+    });
+  }
+
+  /**
+   * This method returns Task that returns value for a single key allowing this strategy to batch operations.
+   * @param key key
+   * @return Task that returns value for a single key allowing this strategy to batch operations
+   */
+  public Task<T> batchable(final K key) {
+    return batchable("taskForKey: " + key.toString(), key);
+  }
+
+  private Task<?> taskForBatch(final G group, final Batch<K, T> batch) {
+    return Task.async(getBatchName(batch, group), ctx -> {
+      final SettablePromise<T> result = Promises.settable();
+      final PromiseListener<T> countDownListener =
+          new CountDownPromiseListener<T>(batch.size(), result, null);
+
+      //one of the batchable tasks becomes a parent of batch task
+      //all other tasks become potential parents
+      final TraceBuilder traceBuilder = ctx.getTraceBuilder();
+      Relationship rel = Relationship.PARENT_OF;
+      for (BatchEntry<T> entry : batch.values()) {
+        for (ShallowTraceBuilder shallowTraceBuilder: entry.getShallowTraceBuilders()) {
+          traceBuilder.addRelationship(rel, shallowTraceBuilder, ctx.getShallowTraceBuilder());
+          rel = Relationship.POTENTIAL_PARENT_OF;
+        }
+        entry.getPromise().addListener(countDownListener);
+      }
+
+      try {
+        if (batch.size() == 1) {
+          Entry<K, BatchEntry<T>> entry = batch.entries().iterator().next();
+          executeSingleton(group, entry.getKey(), entry.getValue());
+        } else {
+          executeBatch(group, batch);
+        }
+      } catch (Throwable t) {
+        batch.failAllRemaining(t);
+      }
+
+      return result;
+    });
+  }
+
+  private Collection<Task<?>> taskForBatches(Collection<Entry<G, Batch<K, T>>> batches) {
+    return batches.stream().map(entry -> taskForBatch(entry.getKey(), entry.getValue())).collect(Collectors.toList());
+  }
+
+  void handleBatch(final PlanContext planContext) {
+    final BatchBuilder<K, T> batchBuilder = _batches.remove(planContext.getId());
+    if (batchBuilder != null) {
+      final Batch<K, T> batch = batchBuilder.build();
+      if (batch.size() > 0) {
+        try {
+          final Map<G, Batch<K, T>> batches = split(batch);
+          if (batches.size() > 0) {
+            taskForBatches(batches.entrySet()).forEach(task -> new ContextImpl(planContext, task).runTask());
+          }
+        } catch (Throwable t) {
+          //we don't care if some of promises have already been completed
+          //all we care is that all remaining promises have been failed
+          batch.failAllRemaining(t);
+        }
+      }
+    }
+  }
+
+  /**
+   * This method will be called for every {@code Batch} that contains at least two elements.
+   * Implementation of this method should make sure that all {@code SettablePromise} contained in the {@code Batch}
+   * will eventually be resolved - typically asynchronously.
```

Comment:
Make this stronger: what happens if you don't resolve all the promises? Is the behaviour undefined or does the Task run indefinitely?